numba/cuda/tests/cudapy/test_libdevice.py

```diff
@@ -0,0 +1,186 @@
+        # as there are in the return type of the libdevice function - one for
+        # scalar-valued returns, or the length of the tuple for tuple-valued
+        # returns.
+        if isinstance(sig.return_type, (types.Tuple, types.UniTuple)):
+            # Start with the parameters for the return values
+            pyargs = ", ".join(['r%d' % i for i in
+                                range(len(sig.return_type))])
+            # Add the parameters for the argument values
+            pyargs += ", " + funcargs
+            # Generate the unpacking of the return value from the libdevice
+            # function into the Python function return values (`r0`, `r1`,
+            # etc.).
+            retvars = ", ".join(['r%d[0]' % i for i in
+                                 range(len(sig.return_type))])
+        else:
+            # Scalar return is a more straightforward case
+            pyargs = "r0, " + funcargs
+            retvars = "r0[0]"
+
+        # Create the string containing the function to compile
+        d = { 'func': apiname,
+              'pyargs': pyargs,
+              'funcargs': funcargs,
+              'retvars': retvars }
+        code = function_template % d
+
+        # Convert the string to a Python function
+        locals = {}
+        exec(code, globals(), locals)
+        pyfunc = locals['pyfunc']
+
+        # Compute the signature for compilation. This mirrors the creation of
+        # arguments to the Python function above.
+        pyargs = [ arg.ty for arg in args if not arg.is_ptr ]
+        if isinstance(sig.return_type, (types.Tuple, types.UniTuple)):
+            pyreturns = [ret[::1] for ret in sig.return_type]
+            pyargs = pyreturns + pyargs
+        else:
+            pyargs.insert(0, sig.return_type[::1])
+
+        ptx, resty = compile_ptx(pyfunc, pyargs)
+
+        # If the function body was discarded by optimization (therefore making
+        # the test a bit weak), there won't be any loading of parameters -
+        # ensure that a load from parameters occurs somewhere in the PTX
+        self.assertIn('ld.param', ptx)
+
+        # Returning the result (through a passed-in array) should also require
+        # a store to global memory, so check for at least one of those too.
+        self.assertIn('st.global', ptx)
+
+    return _test_call_functions
+
+
+@skip_on_cudasim('Compilation to PTX is not supported on cudasim')
+class TestLibdeviceCompilation(unittest.TestCase):
+    """
+    Class for holding all tests of compiling calls to libdevice functions. We
+    generate the actual tests in this class (as opposed to using subTest and
+    one test within this class) because there are a lot of tests, and it makes
+    the test suite appear frozen to test them all as subTests in one test.
+    """
+
+
+for libname in functions:
+    setattr(TestLibdeviceCompilation, 'test_%s' % libname,
+            make_test_call(libname))
```

Comment:
As this is auto-generated, should probably add it to this, largely so it's not forgotten. 
https://github.com/numba/numba/blob/c5788551c582711255dc6ab737e4c76a6e794155/numba/testing/main.py#L337

I don't think it'll have any impact but want to avoid having to debug the class of problem that occurs if injects tests aren't acknowledged.

Reply:
I've now added it to `_GENERATED`.